### PR TITLE
[SPARK-26310][SQL] Verify applicability of JSON options

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose additional verification of JSON options. In particular, detection of using `read` options in `write` and vice versa. The verification can be disabled by new SQL config `spark.sql.verifyDataSourceOptions` (`true` by default). This changes should prevent misusing of JSON options, and improve usage experience of built-in datasource.

## How was this patch tested?

Added new test to `JacksonGeneratorSuite` to check how `in read` option is handled if it is passed `in write`. Also the test checks new SQL config can disable the verification.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@gatorsmile Please, have a look at this PR.

**Comment by MaxGekk** [comment]
@cloud-fan What do you think of the PR?

**Comment by MaxGekk** [comment]
@gatorsmile Should I close the PR or will continue with this approach?

**Comment by MaxGekk** [comment]
@gatorsmile Could you look at the PR, please.

**Comment by MaxGekk** [comment]
@HyukjinKwon What do you think of this PR?

**Review by HyukjinKwon — COMMENTED** [review]
I don't object to the change. My guts say this should be more generalised but on the other hand it looks an overkill. The configuration also looks an overkill - we could warn instead but on the other hand I get why it's needed.

Also, I'm not still sure about `..OptionsRead` and `..OptionsWrite` but I don't object as I said.

**Comment by MaxGekk** [comment]
@HyukjinKwon Thank you for looking at this PR. I proposed the changes because I saw a few times already when our users tried to apply datasource specific options in wrong way. Maybe this is because of Spark's documentation is not clear enough, and we should point out more precisely when particular option can be applicable. ... but I believe it makes sense to solve the problem in code.

> ... more generalised but on the other hand it looks an overkill

I considered more generalized approach like creating a trait for `JSONOptions`, `CSVOptions`, `TextOptions`, `AvroOptions` and etc. like. The trait could contains applicable options for each direction (`in read`/`in write`). In this way, a caller could ask explicitly (or implicitly from the constructor) to verify passed options.

> The configuration also looks an overkill ...

It was introduced only because of behavior change.

> ... we could warn instead but on the other hand I get why it's needed

Logs are not always available to users or can be disabled. And who among users read logs ;-) . I prefer to throw an exception because it can safe user's time in debugging/trouble shooting. I cannot imagine a situation when in write options are passed as read options in purpose, and silent behavior could be expected.

**Comment by MaxGekk** [comment]
@gatorsmile Could you look at this PR, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1635,6 +1635,14 @@ object SQLConf {
       "java.time.* packages are used for the same purpose.")
     .booleanConf
     .createWithDefault(false)
+
+  val VERIFY_DATASOURCE_OPTIONS = buildConf("spark.sql.verifyDataSourceOptions")
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Do we need a config for this? seems more like a bug fix, even if it's a behavior change, and we can be stricter about validation in Spark 3.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I would remove the config but @gatorsmile asked to add it.

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: OK, I can't say I feel too strongly about it, but when would this be set to `false` @gatorsmile ? this is a case where the user is already setting useless options right, so they can a) somehow discover this flag or b) just remove the options. b) seems much more likely to happen. Is the worry about legacy code or libraries that can't be easily modified? those already may not work on Spark 3.

**Comment by MaxGekk** [comment]
I am closing this since I don't see enthusiasm about the changes.

**Pull request closed without merging**